orderer/common/follower/follower_chain.go

```diff
@@ -423,9 +413,130 @@ func (c *Chain) pullUpToJoin() error {
 }
 
 // pullAfterJoin pulls blocks continuously, inspecting the fetched config blocks for membership.
+// On every config block, it renews the BlockPuller, to take in the new configuration.
 // It will exit with 'nil' if it detects a config block that indicates the orderer is a member of the cluster.
 // It checks whether the chain was stopped between blocks.
 func (c *Chain) pullAfterJoin() error {
-	//TODO
-	return errors.New("not implemented yet: pull after join block")
+	var err error
+
+	if c.lastConfig == nil {
+		err = c.loadLastConfig()
+		if err != nil {
+			return errors.Wrap(err, "failed to load last config block")
+		}
+	}
+	c.updateStatusReport()
+	c.logger.Debugf("last config block: %d", c.lastConfig.Header.Number)
+
+	c.blockPuller, err = c.createBlockPullerFunc()
+	if err != nil {
+		return errors.Wrap(err, "error creating block puller")
+	}
+	defer c.blockPuller.Close()
+
+	var isMember bool
+	isMember, err = c.clusterConsenter.IsChannelMember(c.lastConfig)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine channel membership from last config")
+	}
+
+	for !isMember {
+		endpoint, latestHeight, err := cluster.LatestHeightAndEndpoint(c.blockPuller)
+		if err != nil {
+			return errors.Wrap(err, "failed to get latest height and endpoint")
+		}
+		c.logger.Debugf("Orderer endpoint %s has the biggest ledger height: %d", endpoint, latestHeight)
+
+		if latestHeight <= c.ledgerResources.Height() {
+			c.logger.Debugf("My height: %d, latest height: %d; going to wait %v for latest height to grow",
+				c.ledgerResources.Height(), latestHeight, c.retryInterval)
+			select {
+			case <-c.stopChan:
+				return errors.New("stopped while waiting for latest height to grow")
+			case <-c.timeAfter(c.retryInterval):
+				c.increaseRetryInterval() // Exponential back-off, to avoid calling LatestHeightAndEndpoint too often.
+				continue
+			}
+		}
+
+		isMember, err = c.pullUntilLatestOrMembershipConfig(latestHeight)
+		if err != nil {
+			return errors.Wrap(err, "failed to pull until latest height or membership config")
+		}
+		c.decreaseRetryInterval() // After a successful pull, we slowly decrease the retry interval.
+	}
+
+	return nil
+}
+
+// pullUntilLatestOrMembershipConfig is given a height-target and exits without an error when:
+// (a) it reaches that target, returning "false"; or
+// (b) it encounters a config block that indicates channel membership, returning "true".
+func (c *Chain) pullUntilLatestOrMembershipConfig(latestHeight uint64) (bool, error) {
```

Comment:
Why are we pulling only until a membership config? Why not simply until a target block height?  Obviously the membership could change along the way, but why stop replicating?  Replication happens no matter what before we can begin consenting.

Reply:
Done